ruma-events-macros/src/event_enum.rs

```diff
@@ -8,79 +8,72 @@ use syn::{
+            | (EventKind::State(_), EventKindVariation::Stub)
+            | (EventKind::State(_), EventKindVariation::Redacted)
+            | (EventKind::State(_), EventKindVariation::RedactedStub)
+    )
+}
 
-// Arrays of event enum names grouped by a field they share in common.
-const ROOM_EVENT_KIND: &[&str] = &[
-    ANY_MESSAGE_EVENT,
-    ANY_SYNC_MESSAGE_EVENT,
-    ANY_STATE_EVENT,
-    ANY_SYNC_STATE_EVENT,
-    REDACTED_MESSAGE_EVENT,
-    REDACTED_STATE_EVENT,
-    REDACTED_SYNC_MESSAGE_EVENT,
-    REDACTED_SYNC_STATE_EVENT,
-];
+fn room_id_kind(name: &EventKind, var: &EventKindVariation) -> bool {
+    matches!(
+        (name, var),
+        (EventKind::Message(_), EventKindVariation::Full)
+            | (EventKind::Message(_), EventKindVariation::Redacted)
+            | (EventKind::State(_), EventKindVariation::Full)
+            | (EventKind::State(_), EventKindVariation::Redacted)
+            | (EventKind::Ephemeral(_), EventKindVariation::Full)
+    )
+}
 
-const ROOM_ID_KIND: &[&str] = &[
-    ANY_MESSAGE_EVENT,
-    ANY_STATE_EVENT,
-    ANY_EPHEMERAL_EVENT,
-    REDACTED_STATE_EVENT,
-    REDACTED_MESSAGE_EVENT,
-];
+fn event_id_kind(name: &EventKind, var: &EventKindVariation) -> bool {
+    matches!(
+        (name, var),
+        (EventKind::Message(_),EventKindVariation::Full)
+            | (EventKind::Message(_), EventKindVariation::Stub)
+            | (EventKind::Message(_), EventKindVariation::Redacted)
+            | (EventKind::Message(_), EventKindVariation::RedactedStub)
+            | (EventKind::State(_), EventKindVariation::Full)
+            | (EventKind::State(_), EventKindVariation::Stub)
+            | (EventKind::State(_), EventKindVariation::Redacted)
+            | (EventKind::State(_), EventKindVariation::RedactedStub)
+    )
+}
 
-const EVENT_ID_KIND: &[&str] = &[
-    ANY_MESSAGE_EVENT,
-    ANY_SYNC_MESSAGE_EVENT,
-    ANY_STATE_EVENT,
-    ANY_SYNC_STATE_EVENT,
-    REDACTED_SYNC_STATE_EVENT,
-    REDACTED_SYNC_MESSAGE_EVENT,
-    REDACTED_STATE_EVENT,
-    REDACTED_MESSAGE_EVENT,
-];
+fn sender_id_kind(name: &EventKind, _var: &EventKindVariation) -> bool {
+    matches!(name, EventKind::Message(_) | EventKind::State(_) | EventKind::ToDevice(_))
+}
 
-const SENDER_KIND: &[&str] = &[
-    ANY_MESSAGE_EVENT,
-    ANY_STATE_EVENT,
-    ANY_SYNC_STATE_EVENT,
-    ANY_TO_DEVICE_EVENT,
-    ANY_SYNC_MESSAGE_EVENT,
-    ANY_STRIPPED_STATE_EVENT,
-    REDACTED_MESSAGE_EVENT,
-    REDACTED_STATE_EVENT,
-    REDACTED_STRIPPED_STATE_EVENT,
-    REDACTED_SYNC_MESSAGE_EVENT,
-    REDACTED_SYNC_STATE_EVENT,
-];
+fn prev_content_kind(name: &EventKind, var: &EventKindVariation) -> bool {
+    matches!(
+        (name, var),
+        (EventKind::State(_), EventKindVariation::Full) | (EventKind::State(_), EventKindVariation::Stub)
+    )
+}
 
-const PREV_CONTENT_KIND: &[&str] = &[ANY_STATE_EVENT, ANY_SYNC_STATE_EVENT];
+fn state_key_kind(name: &EventKind, _var: &EventKindVariation) -> bool {
```

Comment:
You could inline a bunch of these that are only used once, as closures. (or is there some specific reason you didn't?)

Reply:
Size and readablility but now I think they may be sort enough. I'll commit what I have then inline them so we have a before and after.